keps/1224-lending-limit/README.md

```diff
@@ -0,0 +1,144 @@
+
+### User Stories (Optional)
+
+#### Story 1
+
+In order to ensure the full utilization of resources, we generally set `BorrowingLimit` to max, but this may cause a `ClusterQueue` to run out of its all resources, and make the new incoming job slow to response for the slow preemption. This could be worse in a competitive cluster, jobs will borrow resources and be reclaimed over and over.
+
+So we want to reserve some resources for a `ClusterQueue`, so that any incoming jobs in the  `ClusterQueue` can get admitted immediately.
+
+### Risks and Mitigations
+
+With both BorrowingLimit and LendingLimit configured, one clusterQueue may not be able to borrow up to the limit just because we reserved the lending limit quota of resource. To aviod this, we can leave the BorrowingLimit unset to borrow as mush as we can.
+
+## Design Details
+
+### Kueue LendingLimit API
+
+Modify ResourceQuota API object:
+
+```go
+type ResourceQuota struct {
+	[...]
+
+	// lendingLimit is the maximum amount of quota for the [flavor, resource]
+	// combination that this ClusterQueue is allowed to lend the unused
+	// quota to other ClusterQueues in the same cohort.
+	// In total, at a given time, ClusterQueue can reserve
+	// a quantity of quota equals to nominalQuota - lendingLimit.
+	// If null, it means that there is no lending limit.
+	// If not null, it must be non-negative.
+	// lendingLimit must be null if spec.cohort is empty.
+	// +optional
+	LendingLimit *resource.Quantity `json:"lendingLimit,omitempty"`
+}
+```
+
+Also `Lended` is added to `ResourceUsage`.
+
+```go
+type ResourceUsage struct {
+	[...]
+
+    // Lended is a quantity of quota that is lended to other ClusterQueues in the cohort.
+    Lended resource.Quantity `json:"lended,omitempty"`
+}
+```
+
+
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+None.
+
+#### Unit Tests
+
+- `pkg/cache`: `2023-11-15` - `86.5%`
+- `pkg/controller/core/`: `2023-11-15` - `16.5%`
+- `pkg/metrics/`: `2023-11-15` - `45.7%`
+- `pkg/scheduler/`: `2023-11-15` - `80.6%`
+
+#### Integration tests
+
+<!--
+Describe what tests will be added to ensure proper quality of the enhancement.
+
+After the implementation PR is merged, add the names of the tests here.
+-->
+
+- No new workloads can be admitted when the `LendingLimit` greater than `NominalQuota` or less than `0`.
+- In a cohort with 2 ClusterQueues a, b and single ResourceFlavor:
+  - When cq-b's LendingLimit set:
+    - When cq-a's BorrowingLimit unset, cq-a can borrow as much as `cq-b's LendingLimit`.
+    - When cq-a's BorrowingLimit set, cq-a can borrow as much as `min(cq-b's LendingLimit, cq-a's BorrowingLimit)`.
+- In a ClusterQueue with 2 ResourceFlavors a, b:
+  - When rf-b's LendingLimit set, and FlavorFungibility set to `WhenCanBorrow: Borrow`:
```

Comment:
this is the default fungibility policy.

Is the feature really impacted by different policies?

Reply:
Just in case we missed something, this can be intersected with borrowingLimit on exist testcases. I'm ok to miss this case.